src/core/common/instance.cpp

```diff
@@ -116,21 +120,33 @@ Instance &Instance::InitSingle(void)
 {
     Instance *instance = &Get();
 
+#if !OPENTHREAD_CONFIG_INSTANCE_ON_HEAP
     VerifyOrExit(!instance->mIsInitialized);
 
     instance = new (&gInstanceRaw) Instance();
 
     instance->AfterInit();
+#else
+    VerifyOrExit(instance == NULL);
+
+    gInstanceRaw = new Instance();
+
+    instance = gInstanceRaw;
+
+    instance->AfterInit();
+#endif
 
 exit:
     return *instance;
 }
 
 Instance &Instance::Get(void)
 {
-    void *instance = &gInstanceRaw;
-
-    return *static_cast<Instance *>(instance);
+#if !OPENTHREAD_CONFIG_INSTANCE_ON_HEAP
+    return *static_cast<Instance *>((void*)&gInstanceRaw);
+#else
+    return *gInstanceRaw;
```

Comment:
Should able to use same implementation for both case when `gInstanceRaw` is defined to `uint64_t *`.